guides/building-a-middleware-from-scratch.md

```diff
@@ -0,0 +1,693 @@
+    // Variant used if the inner service produced an error
+    Service(Error),
+}
+```
+
+While this seems ideal on the surface as we're not losing any type information
+and can use `match` to get at the exact error, the approach has three issues:
+
+1. In practice its common to nest lots of middleware. That would make the final
+   error enum very large. Its not unlikely to look something like
+   `BufferError<RateLimitError<TimeoutError<MyError>>>`. Pattern matching on
+   such a type (to for example determine if the error is retry-able) is very
+   tedious.
+2. If we change the order our middleware are applied in we also change the final
+   error type meaning we have to update our pattern matches.
+3. There is also the possibility of the final error type being very large and
+   taking up a significant amount of space on the stack.
+
+With this we're left with option 1 which is to convert the inner service error
+into a boxed trait object like `Box<dyn std::error::Error + Send + Sync>`. That
+means we can combine multiple errors type into one. That has the following
+advantages:
+
+1. Our error handling is less fragile since changing the order middleware are
+   applied in wont change the final error type.
+2. The error type now has a constant size regardless how many middleware we've
+   applied.
+3. Extracting the error no longer requires a big `match` but can instead be done
+   with `error.downcast_ref::<Timeout>()`.
+
+However it also has the following downsides:
+
+1. As we're using dynamic downcasting the compiler can no longer guarantee that
+   we're exhaustively checking for every possible error type.
+2. Creating an error now requires an allocation. In practice we expect errors to
+   be infrequent and therefore this shouldn't be a problem.
+
+Which option you prefer is a matter of personal preference. Both have their
+advantages and disadvantages. However the pattern that we've decided to use in
+Tower is boxed trait objects. You can find the original discussion
+[here](https://github.com/tower-rs/tower/issues/131).
+
+For our `Timeout` middleware that means we need to create a struct that
+implements `std::error::Error` such that we can convert it into a `Box<dyn
+std::error::Error + Send + Sync>`. We also have to require that the inner
+service's error type implements `Into<Box<dyn std::error::Error + Send +
+Sync>>`. Luckily most errors automatically satisfies that so it wont require
+users to write any additional code.
+
+The code for our error type looks like this:
+
+```rust
+use std::fmt;
+
+#[derive(Debug, Default)]
+pub struct TimeoutError(());
+
+impl fmt::Display for TimeoutError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.pad("request timed out")
+    }
+}
+
+impl std::error::Error for TimeoutError {}
+```
+
+We add a private field to `TimeoutError` such that users outside of Tower cannot
+construct their own `TimeoutError`. They can only be obtained through our
+middleware.
```

Comment:
I think the "modern" way of doing this is with `#[non_exhaustive]`.

Reply:
That doesn't seem to work for empty structs https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2ccc7ef1027273fdb59abe545aa20512

Reply:
`#[non_exhaustive]` is only enforced across crate boundaries :)

Reply:
Ah yes of course! Cool. Didn't know 😊